src/node.rs

```diff
@@ -0,0 +1,219 @@
+use crate::helper::Identifier;
+use crate::kbucket::KbucketTable;
+use std::collections::HashMap;
+use std::net::{Ipv4Addr, SocketAddrV4};
+use tokio::io;
+use tokio::net::UdpSocket;
+
+#[derive(Debug)]
+pub enum FindNodeResult {
+    Found(Option<TableRecord>),
+    NotFound(Vec<Option<TableRecord>>),
+}
+
+#[derive(Debug)]
+pub enum Search {
+    Success(usize, usize),
+    Failure(usize, usize),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct TableRecord {
+    pub node_id: Identifier,
+    pub ip_address: Ipv4Addr,
+    pub udp_port: u16,
+    pub socket_addr: SocketAddrV4,
+}
+
+// The main Kademlia client struct.
+// Provides user-level API for performing querie and interacting with the underlying service.
+// #[derive(Clone, Copy, Debug, PartialEq)]
+#[derive(Debug, PartialEq)]
+pub struct Node {
+    pub node_id: Identifier,
+    pub table_record: TableRecord,
+    pub table: KbucketTable,
+    pub store: HashMap<Vec<u8>, Vec<u8>>, // Same storage as Portal network to store samples
+}
+
+impl Node {
+    pub fn new(node_id: Identifier, table_record: TableRecord) -> Self {
+        Self {
+            node_id,
+            table_record,
+            table: KbucketTable::new(node_id),
+            store: Default::default(),
+        }
+    }
+
+    // Protocol's RPCs:
+    // ---------------------------------------------------------------------------------------------------
+    /// "The most important procedure a Kademlia participant must perform is to locate
+    /// the k closest nodes to some given node ID"
+    ///     - Kademlia Paper
+    ///
+    /// Recieves an id request and returns node information on nodes within
+    /// *its closest bucket* (instead of k-closest nodes) to that id.
+    pub fn find_node(&mut self, node_id: &Identifier) -> FindNodeResult {
+        match KbucketTable::search_table(&self.table, node_id) {
+            Search::Success(bucket_index, column_index) => {
+                let bucket = self.table.buckets[bucket_index];
+                FindNodeResult::Found(bucket[column_index])
+            }
+            Search::Failure(bucket_index, column_index) => {
+                let bucket = self.table.buckets[bucket_index];
+                let mut known_nodes = Vec::new();
+
+                for node in bucket.iter() {
+                    if node.is_some() {
+                        // Should I be dereferencing the node to send to others?  Or copy the node to share?
+                        known_nodes.push(*node)
+                    }
+                }
+                FindNodeResult::NotFound(known_nodes)
+            }
```

Comment:
I think you'll be happier if you let everything pull its own weight wrt abstractions

so for example, what does this method do? it looks like it will search the table for some `node_id` and return some nodes from the table

the bits about the table self-organizing along the way ("insert node if missing", "return these other nodes if we don't have this exact one") is a detail that the table should specialize in

to drive the point home, code in this file shouldn't be touching the `buckets` at all; the node has a kademlia table, it doesn't know or care about buckets (afaik)

you do some of this the tests below as well

but I think you'll find this easier to grow/maintain if you keep it as simple as possible and one way to do this is to make good abstractions and then leverage them

Reply:
So a successful search_table( ) should return the node's table record instead of indexes for access to the node elsewhere?